gcloud/storage/blob.py

```diff
@@ -375,60 +378,43 @@ def upload_from_file(self, file_obj, rewind=False, size=None,
-        }
-
-        upload = transfer.Upload(file_obj, content_type, total_bytes,
-                                 auto_transfer=False,
-                                 chunksize=self.chunk_size)
-
-        url_builder = _UrlBuilder(bucket_name=self.bucket.name,
-                                  object_name=self.name)
-        upload_config = _UploadConfig()
-
-        # Temporary URL, until we know simple vs. resumable.
-        base_url = connection.API_BASE_URL + '/upload'
-        upload_url = connection.build_api_url(api_base_url=base_url,
-                                              path=self.bucket.path + '/o')
-
-        # Use apitools 'Upload' facility.
-        request = http_wrapper.Request(upload_url, 'POST', headers)
-
-        upload.ConfigureRequest(upload_config, request, url_builder)
-        query_params = url_builder.query_params
-        base_url = connection.API_BASE_URL + '/upload'
-        request.url = connection.build_api_url(api_base_url=base_url,
-                                               path=self.bucket.path + '/o',
-                                               query_params=query_params)
-        upload.InitializeUpload(request, connection.http)
 
-        # Should we be passing callbacks through from caller?  We can't
-        # pass them as None, because apitools wants to print to the console
-        # by default.
-        if upload.strategy == transfer.RESUMABLE_UPLOAD:
-            http_response = upload.StreamInChunks(
-                callback=lambda *args: None,
-                finish_callback=lambda *args: None)
-        else:
-            http_response = http_wrapper.MakeRequest(connection.http, request,
-                                                     retries=num_retries)
-        response_content = http_response.content
-        if not isinstance(response_content,
-                          six.string_types):  # pragma: NO COVER  Python3
-            response_content = response_content.decode('utf-8')
-        self._set_properties(json.loads(response_content))
+        content_found = self._properties.pop('contentType', None)
+        content_type = (content_type or content_found or
+                        'application/octet-stream')
+        insert_request = storage_v1_messages.StorageObjectsInsertRequest(
+            **self._properties)
+        # Put it back so as note to break the _properties.
+        self._properties['contentType'] = content_found
+
+        insert_request.bucket = self.bucket.name
+        insert_request.name = self.name
+
+        content_type = (content_type or self._properties.get('contentType') or
+                        'application/octet-stream')
+        upload = transfer.Upload.FromStream(file_obj, content_type,
+                                            total_bytes,
+                                            chunksize=self.chunk_size)
+        # @craigcitro, can I put `callback` and `finish_callback` here?
```

Comment:
just sent out a PR to simplify this one.

ultimately, i think you'll *want* callbacks (for checksumming), but for now let's make the easy case easy.